### PR TITLE
Get RTD from ride via a struct method, part 1

### PR DESCRIPTION
**Pull request by Gymnasiast**
The basic idea of this refactor is that the code will retrieve an RTD from a ride and use that for determining ride-related things. This means that ride types will behave in the same way that ride entries and other objects do already.

Access to the RideTypeDescriptors array should eventually be refactored out entirely - code should either use the Ride struct method or use a function like `get_ride_type_descriptor(ObjectEntryIndex index);` to retrieve a specific RTD.

### CODE REVIEW

File: src/openrct2/world/Footpath.cpp [hunk 1]
```diff
@@ -800,9 +800,15 @@ static void loc_6A6D7E(
                     if (initialTileElementPos.z == tileElement->GetBaseZ())
                     {
                         auto ride = get_ride(tileElement->AsTrack()->GetRideIndex());
-                        if (ride == nullptr || !ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_FLAT_RIDE))
+                        if (ride == nullptr)
                         {
-                            continue;
+                            return;
```
Comment (Gymnasiast) [inline src/openrct2/world/Footpath.cpp:805]: revert

File: src/openrct2/world/Footpath.cpp [hunk 2]
```diff
@@ -800,9 +800,15 @@ static void loc_6A6D7E(
...
+                        }
+
+                        const auto& rtd = ride->GetRideTypeDescriptor();
+                        if (!rtd->HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
+                        {
+                            return;
```
Comment (Gymnasiast) [inline src/openrct2/world/Footpath.cpp:811]: continue;

File: test/testpaint/Compat.cpp [hunk 3]
```diff
@@ -844,3 +844,8 @@ void ride_ratings_calculate_lim_launched_roller_coaster([[maybe_unused]] Ride* r
 void ride_ratings_calculate_drink_stall([[maybe_unused]] Ride* ride)
 {
 }
+
+const RideTypeDescriptor& Ride::GetRideTypeDescriptor() const
```
Comment (Gymnasiast) [inline test/testpaint/Compat.cpp:848]:
```suggestion
const RideTypeDescriptor* Ride::GetRideTypeDescriptor() const
```

File: src/openrct2/ride/Ride.cpp [hunk 4]
```diff
@@ -7761,10 +7768,10 @@ uint64_t Ride::GetAvailableModes() const
...
 
-    return GetRideTypeDescriptor().RideModes;
+    return GetRideTypeDescriptor()->RideModes;
 }
 
-const RideTypeDescriptor& Ride::GetRideTypeDescriptor() const
+const RideTypeDescriptor* Ride::GetRideTypeDescriptor() const
 {
-    return RideTypeDescriptors[type];
+    return &RideTypeDescriptors[type];
```
Comment (tupaschoal) [inline src/openrct2/ride/Ride.cpp:7776]: I find this change a bit weird. How can this be a `nullptr`? Do you not initialize the RTD array entirely?

Besides that, you're only checking for the return to be valid on a couple of places.

File: src/openrct2/ride/Ride.cpp [hunk 5]
```diff
@@ -1409,6 +1412,8 @@ void ride_construction_set_default_next_piece()
     if (ride == nullptr)
         return;
 
+    const auto& rtd = ride->GetRideTypeDescriptor();
```
Comment (tupaschoal) [inline src/openrct2/ride/Ride.cpp:1415]: No need for `&` if it's a pointer (applies to multiple other places).

You might not need to change this based on my other comment's outcome.

File: src/openrct2/ride/Ride.cpp (hunk 4 above)
Comment (duncanspumpkin) [reply src/openrct2/ride/Ride.cpp:7776]: Agreed this should return a `const RideTypeDescriptor& ` as there will be no situation where this is nullptr.

Comment (Gymnasiast) [reply src/openrct2/ride/Ride.cpp:7776]: There *will*, once ride types are turned into objects.

Comment (tupaschoal) [reply src/openrct2/ride/Ride.cpp:7776]: Hmm, then what about returning an optional?

Comment (Gymnasiast) [reply src/openrct2/ride/Ride.cpp:7776]: I went for consistency with the other object entry functions (e.g. ride entry, footpath entry, wall entry, ...).

**Comment by Gymnasiast** [comment]
I have updated the PR to use a reference instead of a pointer, at least for now.

File: src/openrct2-ui/windows/Ride.cpp [hunk 6]
```diff
@@ -1048,6 +1048,7 @@ static void window_ride_draw_tab_vehicle(rct_drawpixelinfo* dpi, rct_window* w)
         if (rideEntry == nullptr)
             return;
 
+        const auto& rtd = ride->GetRideTypeDescriptor();
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Ride.cpp:1051]: Move this closer to the if below (or do it directly on the if to reduce the scope of the variable)

File: src/openrct2-ui/windows/Ride.cpp [hunk 7]
```diff
@@ -1060,7 +1061,7 @@ static void window_ride_draw_tab_vehicle(rct_drawpixelinfo* dpi, rct_window* w)
         }
 
         // For any suspended rides, move image higher in the vehicle tab on the rides window
-        if (RideTypeDescriptors[ride->type].HasFlag(RIDE_TYPE_FLAG_IS_SUSPENDED))
+        if (rtd.HasFlag(RIDE_TYPE_FLAG_IS_SUSPENDED))
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/Ride.cpp:1064]:
```
        if (const auto& rtd = ride->GetRideTypeDescriptor(); rtd.HasFlag(RIDE_TYPE_FLAG_IS_SUSPENDED))
```
or 
```
        if (ride->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_IS_SUSPENDED))
```

File: src/openrct2-ui/windows/TileInspector.cpp [hunk 8]
```diff
@@ -1884,7 +1884,8 @@ static void window_tile_inspector_paint(rct_window* w, rct_drawpixelinfo* dpi)
                 auto ride = get_ride(rideId);
                 if (ride != nullptr)
                 {
-                    auto rideType = RideTypeDescriptors[ride->type].Naming.Name;
+                    const auto& rtd = ride->GetRideTypeDescriptor();
+                    auto rideType = rtd.Naming.Name;
```
Comment (tupaschoal) [inline src/openrct2-ui/windows/TileInspector.cpp:1888]:
```suggestion
                    auto rideType = ride->GetRideTypeDescriptor().Naming.Name;
```

File: src/openrct2/actions/LandSetHeightAction.hpp [hunk 9]
```diff
@@ -298,7 +298,8 @@ DEFINE_GAME_ACTION(LandSetHeightAction, GAME_COMMAND_SET_LAND_HEIGHT, GameAction
                         int32_t maxHeight = rideEntry->max_height;
                         if (maxHeight == 0)
                         {
-                            maxHeight = RideTypeDescriptors[ride->type].Heights.MaxHeight;
+                            const auto& rtd = ride->GetRideTypeDescriptor();
+                            maxHeight = rtd.Heights.MaxHeight;
```
Comment (tupaschoal) [inline src/openrct2/actions/LandSetHeightAction.hpp:302]:
```suggestion
                            maxHeight = ride->GetRideTypeDescriptor().Heights.MaxHeight;
```

File: src/openrct2/world/Footpath.cpp [hunk 10]
```diff
@@ -875,7 +875,13 @@ static void loc_6A6D7E(
...
                     {
                         auto ride = get_ride(tileElement->AsTrack()->GetRideIndex());
-                        if (ride == nullptr || !ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_FLAT_RIDE))
+                        if (ride == nullptr)
+                        {
+                            continue;
+                        }
+
+                        const auto& rtd = ride->GetRideTypeDescriptor();
+                        if (!rtd.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```
Comment (tupaschoal) [inline src/openrct2/world/Footpath.cpp:884]:
```suggestion
                        if (ride == nullptr || !ride->GetRideTypeDescriptor(),HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```

File: src/openrct2/world/Footpath.cpp [hunk 11]
```diff
@@ -950,10 +956,17 @@ static void loc_6A6C85(
...
     {
         auto ride = get_ride(tileElementPos.element->AsTrack()->GetRideIndex());
-        if (ride == nullptr || !ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_FLAT_RIDE))
+        if (ride == nullptr)
         {
             return;
         }
+
+        const auto& rtd = ride->GetRideTypeDescriptor();
+        if (!rtd.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```
Comment (tupaschoal) [inline src/openrct2/world/Footpath.cpp:965]: Maybe merge it back into the if above

File: src/openrct2/world/Footpath.cpp [hunk 12]
```diff
@@ -2031,7 +2044,8 @@ bool tile_element_wants_path_connection_towards(const TileCoordsXYZD& coords, co
                     if (ride == nullptr)
                         continue;
 
-                    if (!ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_FLAT_RIDE))
+                    const auto& rtd = ride->GetRideTypeDescriptor();
+                    if (!rtd.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```
Comment (tupaschoal) [inline src/openrct2/world/Footpath.cpp:2048]:
```suggestion
                    if (!ride->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```

File: src/openrct2/world/Footpath.cpp [hunk 13]
```diff
@@ -2120,7 +2134,11 @@ void footpath_remove_edges_at(const CoordsXY& footpathPos, TileElement* tileElem
     {
         auto rideIndex = tileElement->AsTrack()->GetRideIndex();
         auto ride = get_ride(rideIndex);
-        if (ride == nullptr || !ride_type_has_flag(ride->type, RIDE_TYPE_FLAG_FLAT_RIDE))
+        if (ride == nullptr)
+            return;
+
+        const auto& rtd = ride->GetRideTypeDescriptor();
+        if (!rtd.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```
Comment (tupaschoal) [inline src/openrct2/world/Footpath.cpp:2141]:
```suggestion
        if (ride == nullptr || !ride->GetRideTypeDescriptor(),.HasFlag(RIDE_TYPE_FLAG_FLAT_RIDE))
```

File: src/openrct2/world/Footpath.cpp [hunk 14]
```diff
@@ -2271,5 +2289,6 @@ bool PathElement::IsLevelCrossing(const CoordsXY& coords) const
         return false;
     }
 
-    return RideTypeDescriptors[ride->type].HasFlag(RIDE_TYPE_FLAG_SUPPORTS_LEVEL_CROSSINGS);
+    const auto& rtd = ride->GetRideTypeDescriptor();
+    return rtd.HasFlag(RIDE_TYPE_FLAG_SUPPORTS_LEVEL_CROSSINGS);
```
Comment (tupaschoal) [inline src/openrct2/world/Footpath.cpp:2293]:
```suggestion
    return ride->GetRideTypeDescriptor().HasFlag(RIDE_TYPE_FLAG_SUPPORTS_LEVEL_CROSSINGS);
```

**Review by tupaschoal — COMMENTED** [review]
This seems ok, though a few places have variables that I don't think we need (but might have been introduced due to the pointer nature of the previous iteration)

File: src/openrct2-ui/windows/Ride.cpp (hunk 7 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/Ride.cpp:1064]: Think most people decided that using init statement wasn't preferred.

File: src/openrct2-ui/windows/TileInspector.cpp (hunk 8 above)
Comment (duncanspumpkin) [reply src/openrct2-ui/windows/TileInspector.cpp:1888]: Is rideType really a good name for this surely rideTypeName or typeName would be better. rideType generally means the type used to get the rtd.

**Review by duncanspumpkin — APPROVED** [review]
All looks good.

**Comment by tupaschoal** [comment]
I'd still like to see some of my suggested changes incorporated

File: src/openrct2-ui/windows/Ride.cpp (hunk 7 above)
Comment (Gymnasiast) [reply src/openrct2-ui/windows/Ride.cpp:1064]: @tupaschoal I find both suggestions to have way worse readability.

Comment (tupaschoal) [reply src/openrct2-ui/windows/Ride.cpp:1064]: I don't find the second that bad, but your call...

**Pull request merged**